ga4gh/datamodel/genotype_phenotype.py

```diff
@@ -0,0 +1,467 @@
+            raise exceptions.NotImplementedException(
+               "At least one of [location, drug, disease] must be specified")
+        filters = []
+
+        if (location and isinstance(location, basestring)):
+            filters.append('regex(?location_label, "%s")' % location)
+        if (drug and isinstance(drug, basestring)):
+            filters.append('regex(?drug_label, "%s")' % drug)
+        if (disease and isinstance(disease, basestring)):
+            filters.append('regex(?disease_label, "%s")' % disease)
+
+        location_clause = ""
+        if (location and isinstance(location, dict)):
+            locations = []
+            for id in location['ids']:
+                    locations.append('?location = <%s> ' %
+                                     (id['database'] + id['identifier']))
+            location_clause = "(" + " || ".join(locations) + ")"
+            filters.append(location_clause)
+            location_clause = "?l  faldo:location ?location .\n"
+
+        if (drug and isinstance(drug, dict)):
+            drugs = []
+            for id in drug['ids']:
+                    drugs.append('?drug = <%s> ' %
+                                 (id['database'] + id['identifier']))
+            drugs_clause = "(" + " || ".join(drugs) + ")"
+            filters.append(drugs_clause)
+
+        if (disease and isinstance(disease, dict)):
+            diseases = []
+            for id in disease['ids']:
+                    diseases.append('?disease = <%s> ' %
+                                    (id['database'] + id['identifier']))
+            diseases_clause = "(" + " || ".join(diseases) + ")"
+            filters.append(diseases_clause)
+
+        filter = "FILTER (" + " && ".join(filters) + ")"
+        query = query.replace("%FILTER%", filter)
+        query = query.replace("%LOCATION%", location_clause)
+        return query
+
+    def query(self, filter=''):
+        """
+        This is the 'detail' query
+
+        Return a list of dictionaries.
+        Each dict is an [annotation](http://www.w3.org/ns/oa#Annotation)
+        All keys in the dict are predicates of the annotation.
+        All cells in the dict are predicate->object converted to strings.
+
+        If an annotation has a <http://purl.org/oban/association_has_subject>
+        predicate that class is appended to the annotation dict in the
+        'location' property
+        """
+        #
+        annotation_query = """
+        PREFIX OBAN: <http://purl.org/oban/>
+        PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
+        SELECT distinct *
+          WHERE {
+          ?s    a OBAN:association .
+          ?s ?p ?o .
+          OPTIONAL {?o  rdfs:label ?label .} .
+        %FILTER%
+        }
+        """
+
+        annotation_query = annotation_query.replace("%FILTER%", filter)
+        # print(annotation_query)
+        results = G2PDataset.g.query(annotation_query)
+
+        if (G2PDataset.record_test):
```

Comment:
Not sure about this... what's the intention here? We shouldn't be polluting the production code with test code --- these should be fully separated.